### 0. Imports

In [1]:
%load_ext autoreload
%autoreload 2

# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np
import polars as pl

# Visualizaciones
# -----------------------------------------------------------------------
import seaborn as sns
import matplotlib.pyplot as plt

# Vigilar progreso bucles
# -----------------------------------------------------------------------
from tqdm import tqdm

# Gestionar los warnings
# -----------------------------------------------------------------------
import warnings

# modificar el path
# -----------------------------------------------------------------------
import sys
sys.path.append("..")

# importar funciones de soporte
# -----------------------------------------------------------------------
import src.soporte_eda as se
import src.soporte_preprocesamiento as sp

# evaluar objetos literales
# -----------------------------------------------------------------------
from ast import literal_eval 

# statistics functions
# -----------------------------------------------------------------------
from scipy.stats import pearsonr, spearmanr, pointbiserialr


# 1. Introduction - Madrid rental prices - Cleaning

# 2. Preliminary analysis and data cleaning

## 2.1 Import data

In [3]:
pd.set_option("display.max_columns",40)
rent_df = pd.read_csv("../datos/api_rent_madrid.csv",sep=";")
rent_df.head()

propertyCode  numPhotos  price propertyType operation   size  exterior  \
0      98903350         33  550.0       chalet      rent  371.0     False   
1      91615378         12  750.0         flat      rent   60.0      True   
2      88363329         21  750.0         flat      rent   70.0      True   
3      98621753         28  400.0    penthouse      rent   67.0      True   
4      98222714         22  450.0         flat      rent   89.0     False   

   rooms  bathrooms                          address province  \
0      6          3             Numancia de la Sagra   Toledo   
1      1          1  Calle de Alejandro Villegas, 23   Madrid   
2      2          1               Gregorio izquierdo   Madrid   
3      2          2                     Calle Madrid   Madrid   
4      2          1                     Calle Arroyo   Toledo   

                 municipality country     latitude    longitude  showAddress  \
0        Numancia de la Sagra      es  400.675.955  -38.225.891        False   
1                      Madrid      es  404.646.164  -36.489.179         True   
2  San Sebastián de los Reyes      es  405.500.865  -36.278.085        False   
3       Villamanrique de Tajo      es   40.069.197   -3.241.383        False   
4                       Recas      es  400.536.974  -39.912.159        False   

   distance                                        description hasVideo  \
0     40116  UGENA VILLAGE INMOBILIARIA GESTIONA en exclusi...    False   
1      7037  LOFT DE DISEÑO. Orientación Este. Muy soleado ...    False   
2     16145  Estupendo piso muy iluminado, recientemente re...     True   
3     55041  A estrenar ático con piscina en Calle Madrid, ...    False   
4     47186  REDPISO alquila esta vivienda situada en el mu...    False   

  status newDevelopment  priceByArea  \
0   good          False          1.0   
1   good          False         13.0   
2   good          False         11.0   
3   good          False          6.0   
4   good          False          5.0   

                                        detailedType  \
0  {'typology': 'chalet', 'subTypology': 'indepen...   
1                               {'typology': 'flat'}   
2                               {'typology': 'flat'}   
3   {'typology': 'flat', 'subTypology': 'penthouse'}   
4                               {'typology': 'flat'}   

                                      suggestedTexts hasPlan has3DTour has360  \
0  {'subtitle': 'Numancia de la Sagra', 'title': ...   False     False  False   
1  {'subtitle': 'Canillas, Madrid', 'title': 'Pis...   False     False  False   
2  {'subtitle': 'Centro Urbano, San Sebastián de ...    True     False  False   
3  {'subtitle': 'Villamanrique de Tajo', 'title':...   False     False   True   
4  {'subtitle': 'Recas', 'title': 'Piso en Calle ...   False     False  False   

  hasStaging superTopHighlight topNewDevelopment externalReference floor  \
0      False             False             False               NaN   NaN   
1      False             False             False              AV23     3   
2      False             False             False               NaN    bj   
3      False             False             False             90004     2   
4      False             False             False   RP1882022101719     2   

        district neighborhood hasLift  \
0            NaN          NaN     NaN   
1      Hortaleza     Canillas    True   
2  Centro Urbano          NaN   False   
3            NaN          NaN   False   
4            NaN          NaN    True   

                                        parkingSpace labels  \
0                                                NaN    NaN   
1                                                NaN    NaN   
2                                                NaN    NaN   
3                                                NaN    NaN   
4  {'hasParkingSpace': True, 'isParkingSpaceInclu...    NaN   

  newDevelopmentFinished  
0                    NaN  
1              

## 2.2 Explore dataframe

In [4]:
pd.set_option("display.max_rows",40)
se.exploracion_dataframe(rent_df)

El número de datos es 450 y el de columnas es 38

 ..................... 

Las primeras filas del dataframe son:


propertyCode  numPhotos  price propertyType operation   size  exterior  \
0      98903350         33  550.0       chalet      rent  371.0     False   
1      91615378         12  750.0         flat      rent   60.0      True   

   rooms  bathrooms                          address province  \
0      6          3             Numancia de la Sagra   Toledo   
1      1          1  Calle de Alejandro Villegas, 23   Madrid   

           municipality country     latitude    longitude  showAddress  \
0  Numancia de la Sagra      es  400.675.955  -38.225.891        False   
1                Madrid      es  404.646.164  -36.489.179         True   

   distance                                        description hasVideo  \
0     40116  UGENA VILLAGE INMOBILIARIA GESTIONA en exclusi...    False   
1      7037  LOFT DE DISEÑO. Orientación Este. Muy soleado ...    False   

  status newDevelopment  priceByArea  \
0   good          False          1.0   
1   good          False         13.0   

                                        detailedType  \
0  {'typology': 'chalet', 'subTypology': 'indepen...   
1                               {'typology': 'flat'}   

                                      suggestedTexts hasPlan has3DTour has360  \
0  {'subtitle': 'Numancia de la Sagra', 'title': ...   False     False  False   
1  {'subtitle': 'Canillas, Madrid', 'title': 'Pis...   False     False  False   

  hasStaging superTopHighlight topNewDevelopment externalReference floor  \
0      False             False             False               NaN   NaN   
1      False             False             False              AV23     3   

    district neighborhood hasLift parkingSpace labels newDevelopmentFinished  
0        NaN          NaN     NaN          NaN    NaN                    NaN  
1  Hortaleza     Canillas    True          NaN    NaN                    NaN


 ..................... 

Los tipos de las columnas son:


tipo_dato
propertyCode               int64
numPhotos                  int64
price                    float64
propertyType              object
operation                 object
size                     float64
exterior                    bool
rooms                      int64
bathrooms                  int64
address                   object
province                  object
municipality              object
country                   object
latitude                  object
longitude                 object
showAddress                 bool
distance                   int64
description               object
hasVideo                  object
status                    object
newDevelopment            object
priceByArea              float64
detailedType              object
suggestedTexts            object
hasPlan                   object
has3DTour                 object
has360                    object
hasStaging                object
superTopHighlight         object
topNewDevelopment         object
externalReference         object
floor                     object
district                  object
neighborhood              object
hasLift                   object
parkingSpace              object
labels                    object
newDevelopmentFinished    object


 ..................... 

Los duplicados que tenemos en el conjunto de datos son: 8

 ..................... 

Los nulos que tenemos en el conjunto de datos son:


%_nulos
description              0.222222
hasVideo                 4.000000
status                   4.000000
newDevelopment           4.000000
priceByArea              4.000000
detailedType             4.000000
suggestedTexts           4.000000
hasPlan                  4.000000
has3DTour                4.000000
has360                   4.000000
hasStaging               4.000000
superTopHighlight        4.000000
topNewDevelopment        4.000000
externalReference       23.333333
floor                   18.666667
district                24.888889
neighborhood            61.333333
hasLift                  6.888889
parkingSpace            68.888889
labels                  93.333333
newDevelopmentFinished  96.222222


 ..................... 

Los valores que tenemos para las columnas categóricas son: 
La columna PROPERTYTYPE tiene 6 valores únicos, de los cuales los primeros son:


count   pct
propertyType             
flat            356  79.1
studio           48  10.7
duplex           19   4.2
penthouse        16   3.6
chalet            9   2.0

La columna OPERATION tiene 1 valores únicos, de los cuales los primeros son:


count    pct
operation              
rent         450  100.0

La columna ADDRESS tiene 405 valores únicos, de los cuales los primeros son:


count  pct
address                                   
Calle Valle de Ambles, 2           10  2.2
El Quiñón                           4  0.9
Calle de Sáhara                     3  0.7
Calle de Honduras                   3  0.7
Calle de Navalmoral de la Mata      2  0.4

La columna PROVINCE tiene 5 valores únicos, de los cuales los primeros son:


count   pct
province                
Madrid         358  79.6
Toledo          62  13.8
Guadalajara     21   4.7
Segovia          8   1.8
Ávila            1   0.2

La columna MUNICIPALITY tiene 96 valores únicos, de los cuales los primeros son:


count   pct
municipality                  
Madrid               170  37.8
Alcalá de Henares     23   5.1
Seseña                14   3.1
Torrejón de Ardoz     14   3.1
Parla                 14   3.1

La columna COUNTRY tiene 1 valores únicos, de los cuales los primeros son:


count    pct
country              
es         450  100.0

La columna LATITUDE tiene 433 valores únicos, de los cuales los primeros son:


count  pct
latitude               
404.430.651     10  2.2
403.809.876      2  0.4
402.003.307      2  0.4
405.126.166      2  0.4
400.493.725      2  0.4

La columna LONGITUDE tiene 433 valores únicos, de los cuales los primeros son:


count  pct
longitude              
-34.659.877     10  2.2
-36.338.528      2  0.4
-36.919.489      2  0.4
-33.694.194      2  0.4
-42.105.716      2  0.4

La columna DESCRIPTION tiene 427 valores únicos, de los cuales los primeros son:


count  pct
description                                                   
VIVIENDA NUEVA PARA ALQUILER ¡Estrena hogar con...     10  2.2
Nuevo curso, nuevo hogar ¡con hasta 1 MES GRATI...      2  0.4
InmoRenta                                               2  0.4
LEER TEXTO COMPLETO Apartamento totalmente amue...      2  0.4
PROMOCIÓN: 1 MES DE ALQUILER GRATIS si reservas...      2  0.4

La columna HASVIDEO tiene 2 valores únicos, de los cuales los primeros son:


count   pct
hasVideo             
False       314  69.8
True        118  26.2

La columna STATUS tiene 3 valores únicos, de los cuales los primeros son:


count   pct
status                     
good              413  91.8
newdevelopment     17   3.8
renew               2   0.4

La columna NEWDEVELOPMENT tiene 2 valores únicos, de los cuales los primeros son:


count   pct
newDevelopment             
False             415  92.2
True               17   3.8

La columna DETAILEDTYPE tiene 10 valores únicos, de los cuales los primeros son:


count   pct
detailedType                                                   
{'typology': 'flat'}                                  344  76.4
{'typology': 'flat', 'subTypology': 'studio'}          46  10.2
{'typology': 'flat', 'subTypology': 'penthouse'}       16   3.6
{'typology': 'flat', 'subTypology': 'duplex'}          15   3.3
{'typology': 'chalet', 'subTypology': 'terraced...      5   1.1

La columna SUGGESTEDTEXTS tiene 395 valores únicos, de los cuales los primeros son:


count  pct
suggestedTexts                                                
{'subtitle': 'Suroeste, Torrejón de Ardoz', 'ti...     10  2.2
{'subtitle': 'El Quiñón, Seseña', 'title': 'Piso'}      4  0.9
{'subtitle': 'Los Rosales, Madrid', 'title': 'P...      3  0.7
{'subtitle': 'Las Sedas - El Olivar, Alcalá de ...      3  0.7
{'subtitle': 'Cabañas de la Sagra', 'title': 'P...      2  0.4

La columna HASPLAN tiene 2 valores únicos, de los cuales los primeros son:


count   pct
hasPlan             
False      296  65.8
True       136  30.2

La columna HAS3DTOUR tiene 2 valores únicos, de los cuales los primeros son:


count   pct
has3DTour             
False        426  94.7
True           6   1.3

La columna HAS360 tiene 2 valores únicos, de los cuales los primeros son:


count   pct
has360             
False     345  76.7
True       87  19.3

La columna HASSTAGING tiene 1 valores únicos, de los cuales los primeros son:


count   pct
hasStaging             
False         432  96.0

La columna SUPERTOPHIGHLIGHT tiene 1 valores únicos, de los cuales los primeros son:


count   pct
superTopHighlight             
False                432  96.0

La columna TOPNEWDEVELOPMENT tiene 1 valores únicos, de los cuales los primeros son:


count   pct
topNewDevelopment             
False                432  96.0

La columna EXTERNALREFERENCE tiene 335 valores únicos, de los cuales los primeros son:


count  pct
externalReference            
CG-91792-0001          2  0.4
3,80811E+11            2  0.4
VT02                   2  0.4
VT07                   2  0.4
VT3                    2  0.4

La columna FLOOR tiene 13 valores únicos, de los cuales los primeros son:


count   pct
floor             
1         95  21.1
2         77  17.1
bj        68  15.1
3         60  13.3
4         36   8.0

La columna DISTRICT tiene 96 valores únicos, de los cuales los primeros son:


count   pct
district                       
Centro                 48  10.7
Villaverde             23   5.1
Puente de Vallecas     17   3.8
El Quiñón              13   2.9
Tetuán                 13   2.9

La columna NEIGHBORHOOD tiene 74 valores únicos, de los cuales los primeros son:


count  pct
neighborhood                    
Lavapiés-Embajadores     11  2.4
Villaverde Alto           9  2.0
Valdeacederas             8  1.8
San Diego                 8  1.8
Los Rosales               7  1.6

La columna HASLIFT tiene 2 valores únicos, de los cuales los primeros son:


count   pct
hasLift             
True       234  52.0
False      185  41.1

La columna PARKINGSPACE tiene 5 valores únicos, de los cuales los primeros son:


count   pct
parkingSpace                                                   
{'hasParkingSpace': True, 'isParkingSpaceInclud...    136  30.2
{'hasParkingSpace': True, 'isParkingSpaceInclud...      1   0.2
{'hasParkingSpace': True, 'isParkingSpaceInclud...      1   0.2
{'hasParkingSpace': True, 'isParkingSpaceInclud...      1   0.2
{'hasParkingSpace': True, 'isParkingSpaceInclud...      1   0.2

La columna LABELS tiene 1 valores únicos, de los cuales los primeros son:


count  pct
labels                                                        
[{'name': 'apartamentoType', 'text': 'Apartamen...     30  6.7

La columna NEWDEVELOPMENTFINISHED tiene 2 valores únicos, de los cuales los primeros son:


count  pct
newDevelopmentFinished            
True                       12  2.7
False                       5  1.1

From a quick first data types and column value inspection, we observe:

**Low variability columns**
- operation: A single value column. Should be dropped as it adds no variance.
- Country: A single value column. Should be dropped as it adds no variance.
- Status: The most frequent category has a 91.8% of observations. The second just 3.8% meaning 17 observations, which would leave very little room for an algorithm to have a significant sample for that second category. It is better dropped.
- newDevelopment: Same as Status.
- detailedtype: Most flats (76.4%) do not posses a subTypology. Thus, it is the same as having 76.4% NaN values.
- has3DTour: With just 6 samples for the binary minority class, it brings no information.
- hasStaging: A single value column. Should be dropped as it adds no variance.
- superTopHighlight: A single value column. Should be dropped as it adds no variance.
- topNewDevelopment: A single value column. Should be dropped as it adds no variance.


In [5]:
rent_df.drop(columns=["operation","country","status", "newDevelopment",
                      "detailedType","has3DTour","hasStaging","superTopHighlight",
                      "topNewDevelopment"],inplace=True)




**Duplicates**
- There are 8 row level duplicates in the dataset.


In [6]:
rent_df.drop_duplicates(inplace=True)

- There are other suposed unique values, such as the description or latitude, longitude that are duplicaded, which is worth inspecting:

In [7]:
rent_df[rent_df.duplicated(subset="description",keep=False)].head(4)

propertyCode  numPhotos  price propertyType  size  exterior  rooms  \
53      98708163         13  650.0         flat  70.0      True      2   
55      98715256         13  650.0         flat  70.0      True      2   
78      98932280         20  650.0         flat  55.0      True      1   
85      98931699         20  650.0         flat  55.0      True      1   

    bathrooms       address province municipality     latitude    longitude  \
53          1         jerez   Toledo  Calypo Fado   40.247.822   -4.060.189   
55          1         jerez   Toledo  Calypo Fado  402.449.057  -40.682.374   
78          1  barrio Rejas   Madrid       Madrid  404.415.556  -35.836.227   
85          1  barrio Rejas   Madrid       Madrid  404.414.556  -35.860.227   

    showAddress  distance                                        description  \
53        False     35607  piso azul inmobiliaria alquila en calypo fado,...   
55        False     36358  piso azul inmobiliaria alquila en calypo fado,...   
78        False     10495  LEER TEXTO COMPLETO Apartamento totalmente amu...   
85        False     10296  LEER TEXTO COMPLETO Apartamento totalmente amu...   

   hasVideo  priceByArea                                     suggestedTexts  \
53    False          9.0  {'subtitle': 'Calypo Fado', 'title': 'Piso en ...   
55    False          9.0  {'subtitle': 'Calypo Fado', 'title': 'Piso en ...   
78    False         12.0     {'subtitle': 'Rejas, Madrid', 'title': 'Piso'}   
85    False         12.0     {'subtitle': 'Rejas, Madrid', 'title': 'Piso'}   

   hasPlan has360 externalReference floor  district neighborhood hasLift  \
53   False  False               NaN     1       NaN          NaN   False   
55   False  False               NaN     1       NaN          NaN   False   
78   False  False       11DEY750/11     4  San Blas        Rejas    True   
85   False  False       10DEY750/10     4  San Blas        Rejas    True   

                                         parkingSpace  \
53                                                NaN   
55                                                NaN   
78  {'hasParkingSpace': True, 'isParkingSpaceInclu...   
85  {'hasParkingSpace': True, 'isParkingSpaceInclu...   

                                               labels newDevelopmentFinished  
53                                                NaN                    NaN  
55                                                NaN                    NaN  
78  [{'name': 'apartamentoType', 'text': 'Apartame...                    NaN  
85  [{'name': 'apartamentoType', 'text': 'Apartame...                    NaN

It seems that except for the property code, records sharing the same description are in fact duplicated and should thus be dropped.

In [8]:
rent_df.shape[0]

442

In [9]:
rent_df.drop_duplicates(subset="description",keep="first", inplace=True)

In [10]:
rent_df.shape[0]

428

**Missing values**

Most columns have a 4% of missing values that should be treated. The following have more than 5% but less than 50% and could be interesting to keep in abscence of other errors:
- floor
- district
- hasLift

In this case, as they are of categorical nature, they can be handled straight away by creating a new 'unknown' category in the case of floor and district. In the case of hasLift, it can be inferred that if this is not specified in the listing it is due to a lack of this feature, so it can be set to false.

In [11]:
rent_df["district"].fillna("unknown",inplace=True)
rent_df["hasLift"].fillna(False,inplace=True)

C:\Users\Miguel\AppData\Local\Temp\ipykernel_20764\1949038188.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  rent_df["district"].fillna("unknown",inplace=True)
C:\Users\Miguel\AppData\Local\Temp\ipykernel_20764\1949038188.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For

The following columns should be dropped due to high volume (more than 50%) of missing values:
- neighborhood
- labels
- newDevelopmentFinished

In [12]:
rent_df.drop(columns=["neighborhood","labels","newDevelopmentFinished"],inplace=True)

The column 'parkingSpace' has above 68% of missing values, but this one being almost binary could be kept. Missing values could be kept as 'unknown', replacing a False value, as most non-missing values are True, in case it brings useful information. This is to be contrasted in subsequent steps.

In [13]:
# use pd.categorical as .astype("category") raises errors
rent_df["parkingSpace_included_in_listing"] = np.where(extract_value(rent_df['parkingSpace'], 'isParkingSpaceIncludedInPrice') == True,True,False)
rent_df.drop(columns="parkingSpace",inplace=True)

### Boolean columns values

Explore values from boolean columns:

In [14]:
for column in rent_df.select_dtypes(bool):
    
    display(rent_df[column].value_counts())

exterior
True     347
False     81
Name: count, dtype: int64

showAddress
False    333
True      95
Name: count, dtype: int64

hasLift
True     216
False    212
Name: count, dtype: int64

parkingSpace_included_in_listing
False    307
True     121
Name: count, dtype: int64

A priori, everything is in order.

### Categorical columns values
**High cardinality**
- Description: For obvious reasons. Some characteristics could be extracted, such as wether the rental unit includes furniture or not. Do not drop for now.
- SuggestedTexts: High cardinality, possible does not add any value. Drop.
- externalReference: High cardinality, possible does not add any value as it looks like an id column. Drop.
- address: it quite possibly adds no value. Leave for now in case it can serve the purpose of getting proxy variables if the model needs it.


In [15]:
rent_df.drop(columns=["suggestedTexts","externalReference"], inplace=True)

**Data types and errors**

- Latitude & Longitude: Object data types are incorrect due to typo in the coordinates. Should be float.


In [16]:
rent_df["latitude"] = rent_df["latitude"].str.replace(".","").str[:7].astype(int) / 10**5
rent_df["longitude"] = rent_df["longitude"].str.replace(".","").str[:7].astype(int) / 10**5

- HasVideo: should be boolean.
- hasPlan to has360: Should be boolean.
- hasLift: should be boolean

In [17]:
rent_df[["hasVideo","hasPlan","has360","hasLift"]] = rent_df[["hasVideo","hasPlan","has360","hasLift"]].astype(bool)

- floor: although categorical, should be in numeric format. Assuming "bj" = "bajo", "en" = "entresuelo", "st" = "sotano" and "ss" = "semisotano", then:

In [18]:
floor_replacement_dict = {
    "bj": 0,     # bajo 
    "en": 0.5,   # entresuelo
    "st": -1,    # sótano 
    "ss": -0.5   # semisótano 
}

rent_df["floor"] = rent_df["floor"].replace(floor_replacement_dict).astype("object")

###

Additional concerns like category underrepresentation will be addressed after further EDA and data understanding.

### Numerical columns

For numerical columns, the quickest way to assess value ranges and other statistics is through a descriptive summary:

In [19]:
rent_df.describe().T

count          mean           std           min           25%  \
propertyCode  428.0  9.447580e+07  1.418609e+07  666696.00000  9.752670e+07   
numPhotos     428.0  2.188551e+01  9.735542e+00       0.00000  1.500000e+01   
price         428.0  6.726963e+02  8.499649e+01     340.00000  6.400000e+02   
size          428.0  6.779673e+01  3.636241e+01      20.00000  4.600000e+01   
rooms         428.0  1.693925e+00  9.857346e-01       0.00000  1.000000e+00   
bathrooms     428.0  1.193925e+00  4.298690e-01       1.00000  1.000000e+00   
latitude      428.0  4.030558e+01  1.766794e+00       4.03751  4.030286e+01   
longitude     428.0 -3.682008e+00  3.205519e-01      -4.33449 -3.776485e+00   
distance      428.0  2.229495e+04  1.720615e+04     183.00000  6.991000e+03   
priceByArea   412.0  1.201699e+01  5.421611e+00       1.00000  8.000000e+00   

                       50%           75%           max  
propertyCode  9.866600e+07  9.889510e+07  9.893968e+07  
numPhotos     2.100000e+01  2.800000e+01  7.200000e+01  
price         7.000000e+02  7.400000e+02  7.500000e+02  
size          6.150000e+01  8.000000e+01  3.710000e+02  
rooms         2.000000e+00  2.000000e+00  6.000000e+00  
bathrooms     1.000000e+00  1.000000e+00  3.000000e+00  
latitude      4.040550e+01  4.048334e+01  4.090739e+01  
longitude    -3.696145e+00 -3.604645e+00 -3.577900e-01  
distance      1.963150e+04  3.516600e+04  5.991900e+04  
priceByArea   1.100000e+01  1.500000e+01  3.300000e+01

No presence of outliers is detected judging from the ranges of features like price, size and distance.

PropertyCode should be dropped, as it is just an ID column with no predictive power and no potential use.

In [20]:
rent_df.drop(columns="propertyCode", inplace=True)

Price shows a healthy range of values with no odd numbers (negative, 999 format, etc) within range that could encode missing values.

Rooms and bathroom probably should not be treated as numerical as they probably have very few unique values:


In [21]:
print(f"Rooms has {rent_df.rooms.nunique()} values. Bathrooms has {rent_df.bathrooms.nunique()}.")

# convert to object to be treated as category
rent_df["rooms"] = rent_df["rooms"].astype("object")
rent_df["bathrooms"] = rent_df["bathrooms"].astype("object")

Rooms has 6 values. Bathrooms has 3.



Appart from that, there is the minimum price by area of 1 euro, that looks suspiciously low. Let's check it:

In [22]:
rent_df[["price","size","address","municipality","province","priceByArea"]].sort_values("priceByArea").head(3)

price   size               address            municipality province  \
0    550.0  371.0  Numancia de la Sagra    Numancia de la Sagra   Toledo   
415  750.0  341.0         sexta bandera  Las Ventas de Retamosa   Toledo   
296  600.0  180.0                  Meco                    Meco   Madrid   

     priceByArea  
0            1.0  
415          2.0  
296          3.0

It actually looks like 'priceByArea' has correct values, except from the fact that they should be float values. Recalculating it:

In [23]:
rent_df["priceByArea"] = rent_df["price"] / rent_df["size"]

All other values present healthy ranges. The only feature that raises an eyebrow is distance, as it is not quite clear with respect to what that distance is measured. So let's check to what other feature's values its minimum values correspond:

In [24]:
rent_df[["address","municipality","distance"]].sort_values("distance").head(2)

address municipality  distance
133   Calle Carretas, 31       Madrid       183
154  Calle de Atocha, 51       Madrid       470

The shortest distance corresponds to Calle Carretas, 31. A quick google look up returns that it is close to "Kilometer 0" in Puerta del Sol, and the distance coincides with 183m. 

This distance measure is rather brute, and probably not as informative as the median price of living in certain districts or streets as it informs merely about distance to the center of Madrid. However, let's keep it for now to analyse it via due EDA.

# 2. Export for EDA

Data has been preliminarily explored and cleaned. The resulting dataset will be exported to continue with EDA in `notebooks/1_exploration_v1.ipynb`

In [25]:
rent_df.columns

Index(['numPhotos', 'price', 'propertyType', 'size', 'exterior', 'rooms',
       'bathrooms', 'address', 'province', 'municipality', 'latitude',
       'longitude', 'showAddress', 'distance', 'description', 'hasVideo',
       'priceByArea', 'hasPlan', 'has360', 'floor', 'district', 'hasLift',
       'parkingSpace_included_in_listing'],
      dtype='object')

In [32]:
rent_df_cleaned = rent_df[['numPhotos', 'price', 'propertyType', 'size', 'exterior', 'rooms',
       'bathrooms', 'distance', 'priceByArea', 'hasPlan', 'has360', 'hasVideo', 'description',
       'hasLift', 'parkingSpace_included_in_listing','municipality','province','floor']]

# version for first iteration of preprocessing
rent_df_cleaned.to_csv("../datos/cleaned/rent_df_cleaned_v1.csv")